### DISCUSSION TITLE
Adding GUIslice to PlatformIO Throws a Compile Error

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [Bodmer/TFT_eSPI#1697] [Compatible support libraries]
I'm asking over at GUIslice as well as PlatformIO but I thought I'd circle back here as well just in case someone has come across this.

I had my code working well in PlatformIO using TFT_eSPI but I wanted to experiment with GUIslice.

What I found using one of the GUIslice examples is that it compiles if I add TFT_eSPI as a library dependency but if I add my build flags then the compile fails with this:

```
Compiling .pio\build\esp32dev\FrameworkArduino\MD5Builder.cpp.o
.pio/libdeps/esp32dev/GUIslice/src/GUIslice_drv_tft_espi.cpp: In function 'bool gslc_DrvGetTxtSize(gslc_tsGui*, gslc_tsFont*, const char*, gslc_teTxtFlags, int16_t*, int16_t*, uint16_t*, uint16_t*)':
.pio/libdeps/esp32dev/GUIslice/src/GUIslice_drv_tft_espi.cpp:394:31: error: ISO C++ forbids declaration of 'type name' with no type [-fpermissive]
     m_disp.setFreeFont((const GFXfont *)pFont->pvFont);
Compiling .pio\build\esp32dev\FrameworkArduino\Print.cpp.o
                               ^
.pio/libdeps/esp32dev/GUIslice/src/GUIslice_drv_tft_espi.cpp:394:25: error: expected primary-expression before 'const'
     m_disp.setFreeFont((const GFXfont *)pFont->pvFont);
                         ^
```

When I follow the trail to GUIslice_drv_tft_espi.cpp:394:31 VS Code tells me that GFXfont is undefined but it is defined in [Adafruit-GFX-Library/gfxfont.h at master · adafruit/Adafruit-GFX-Library (github.com)](https://github.com/adafruit/Adafruit-GFX-Library/blob/master/gfxfont.h) which is included in [Adafruit-GFX-Library/Adafruit_GFX.h at master · adafruit/Adafruit-GFX-Library (github.com)](https://github.com/adafruit/Adafruit-GFX-Library/blob/master/Adafruit_GFX.h)

My complete (including some experimenting) platformio.ini is:

```
[platformio]
;src_dir = examples/arduino/ex46_ard_togglebtn
;lib_dir = .

[env]
;lib_ldf_mode = deep ;chain+ ;i also tried deep

[env:esp32dev]
platform = espressif32
;platform = https://github.com/platformio/platform-espressif32.git  ; will force to use the 2.0 board package
board = esp32dev
framework = arduino
lib_deps =
  SPI
  Wire
  adafruit/Adafruit BusIO
  adafruit/Adafruit GFX Library
;  adafruit/Adafruit ILI9341
  impulseadventure/GUIslice
  TFT_eSPI
  adafruit/Adafruit FT6206 Library  ; needed for GUIslice

build_flags =
  -DUSER_CONFIG_LOADED
  -DUSER_CONFIG_INC_FILE
  -DUSER_CONFIG_INC_FNAME=\"../configs/esp-tftespi-default-ft6206.h\"
  # Now TFT_eSPI Build Flags
  -DUSER_SETUP_LOADED=1
  -DBOARD_HAS_PSRAM=1
  -DILI9488_DRIVER=1                           ; Select ILI9488 driver
  -DTFT_WIDTH=320                              ; Set TFT size
  -DTFT_HEIGHT=480
  -DTFT_MISO=19                                ; Define SPI pins
  -DTFT_MOSI=23
  -DTFT_SCLK=18
  -DTFT_CS=5
  -DTFT_DC=14                                  ; Data/Command pin
  -DTFT_RST=13                                 ; Reset pin
  -DTFT_BL=12                                  ; LED backlight control pin
  -DTFT_BACKLIGHT_ON=HIGH                      ; Level to turn on backlight
;  -DTOUCH_CS=21                                ; Not used but needed for compile
  -DLOAD_GLCD=1                                ; Load Fonts
  -DLOAD_FONT2=1          ; Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
  -DLOAD_FONT4=1          ; Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
  -DLOAD_FONT6=1          ; Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
  -DLOAD_FONT7=1          ; Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
  -DLOAD_FONT8=1          ; Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
  -DSMOOTH_FONT=1
  -DSPI_FREQUENCY=27000000                     ; Set SPI frequency
```

So if you cut everything below `# Now TFT_eSPI Build Flags` you'll get a successful compile but if you add those build flags back in I got the odd error listed above.

I know there isn't CUIslice support here but I thought I'd post in case someone has run across something similar and can help me debug this further.

**Comment by GeorgeIoak** [comment]
OK, just stumbled across that if I comment out
`;  -DUSER_SETUP_LOADED=1`
I can get it to compile

So there must be a downside to doing this or since `User_setup_Select.h` doesn't load any configurations by default I think all my build flags are going to work and do everything I need?

**Comment by GeorgeIoak** [comment]
OK, just stumbled across that if I comment out
; -DUSER_SETUP_LOADED=1
I can get it to compile. Since I still have the rest of my build flags I think I’m OK and things will work but I’m a bit confused on why I’m getting that compile error. Unfortunately when I do load it to my board I get stuck in a fault and the board reboots:

As a reminder all of the settings worked with a different project before I added GUIslice so the hardware and settings are known to work.
```
Rebooting...
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:1044
load:0x40078000,len:10124
load:0x40080400,len:5828
entry 0x400806a8
GUIslice version [0.17.0]:
- Init display handler [TFT_eSPI] OK
- Init touch handler [FT6206(I2C)] FAIL
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
Core 1 register dump:
PC      : 0x400d3437  PS      : 0x00060530  A0      : 0x800d2fc1  A1      : 0x3ffb1ef0
A2      : 0x3ffc0324  A3      : 0x3ffc0894  A4      : 0x00000005  A5      : 0x00000000
A6      : 0x3f400b6b  A7      : 0x00000000  A8      : 0x800d28fc  A9      : 0x3ffb1e20
A10     : 0x3ffc0324  A11     : 0x3ffbff50  A12     : 0x00200000  A13     : 0x00000091  
A14     : 0x3ffc0268  A15     : 0x00000000  SAR     : 0x0000001c  EXCCAUSE: 0x0000001c
EXCVADDR: 0x00000010  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0x00000000

ELF file SHA256: 0000000000000000

Backtrace: 0x400d3437:0x3ffb1ef0 0x400d2fbe:0x3ffb1f30 0x400d1602:0x3ffb1f90 0x400d54d5:0x3ffb1fb0 0x40086b05:0x3ffb1fd0        
```

I see `- Init touch handler [FT6206(I2C)] FAIL` which reminds me that my I2C pins are on different ESP32 pins. In my original program I had:
```
#define I2C_Freq 100000
#define I2C_SDA 33
#define I2C_SCL 27

TwoWire I2C_0 = TwoWire(0);
```
and then in `setup()` I had:
```
 I2C_0.begin(I2C_SDA, I2C_SCL, I2C_Freq);

  if (!ts.begin(TOUCH_THRESHOLD, I2C_SDA, I2C_SCL))
  {
    Serial.println("[INFO]: Capacitive touch FAILED to start");
  }
  else
    Serial.println("[INFO]: Capacitive touch started!");
```

But I'm not quite sure how you go about using custom I2C pins with GUIslice

**Comment by Bodmer** [comment]
Looks like you are missing a define:

#define LOAD_GFXFF

So in platformIO you need to add this to the ini file:

-DLOAD_GFXFF=1

**Reply by GeorgeIoak** [reply]
Thanks, Paul over at GUIslice caught that too. I'm still playing around (with too many variations) and make 1 step forward and 2 steps back. I'm hoping to get a better handle on the inner working of PlatformIO and come back to this thread and hopefully post some pointers to others.
Right now I'm still in the trial and error stage with many fails and few successes but it is getting better.

**Reply by Bodmer** [reply]
Looks like GUISlice is being worked on again, so that is good news.
